README.md
Description
===========

Installs and configures
Tomcat version 7.0.33 with OpenAM 10.1.0
OpenAM provides Access Management, Federated SSO. For more information,
[Forgerock - OpenAM](http://www.forgerock.org/openam.html)


Where can this be used ?
============
Helps to setup an cloud indentity for your organization. Developers/Testers can standup an cloud identity to test against.


Requirements
============

Platform:

* Debian, Ubuntu (OpenJDK)


The following Opscode cookbooks are dependencies:

* apache2
* java
* apt
* opendj `(If you wish to have seperate opendj)`

Attributes
==========
[OpenAM Configuration](http://openam.forgerock.org/openam-documentation/openam-doc-source/doc/install-guide/index.html)
* `node["tomcat-openam"]["opendj"]["arg-val"]["baseDN"]` - LDAP BaseDN value, default `dc=example,dc=com`
* `node["tomcat-openam"]["opendj"]["arg-val"]["rootUserDN"]` - LDAP root user name, default `'n=Directory Manager'`
* `node["tomcat-openam"]["opendj"]["arg-val"]["rootUserPassword"]` - LDAP root user password, default `secret12`
* `node["tomcat-openam"]["opendj"]["arg-val"]["ldapPort"]` - LDAP port value, default `1389`

* `node["tomcat-openam"]["java-options"]` - Heap storage limits, default `-Xms256m -Xmx1024m`

* `node["tomcat-openam"]["cfg"]["server-url"]` - OpenAM server url, default `http://#{node["tomcat-openam"]["dns"]}:8080`
* `node["tomcat-openam"]["cfg"]["deployment-uri"]` - OpenAM deployment uri, default `/openam`
* `node["tomcat-openam"]["cfg"]["base-dir"]` - OpenAM base directory, default `/home/ubuntu/openam`
* `node["tomcat-openam"]["cfg"]["locale"]` - Locale for OpenAM, default `en_US`
* `node["tomcat-openam"]["cfg"]["platform-locale"]` - Platform Locale for OpenAM, defualt `en_US`
* `node["tomcat-openam"]["cfg"]["admin-pwd"]` - OpenAM admin password, default `adminp3me`
* `node["tomcat-openam"]["cfg"]["amldapuserpasswd"]` - LDAP user password, default `adminl3me`
* `node["tomcat-openam"]["cfg"]["cookie-domain"]` - Domain for cookie, default `node[:ec2][:public_hostname]`
* `node["tomcat-openam"]["cfg"]["data-store"]` - OpenAM data store, default `embedded`

* `node["tomcat-openam"]["cfg"]["directory"]["ssl"]` - OpenAM SSL directory, default `SIMPLE`
* `node["tomcat-openam"]["cfg"]["directory"]["server"]` - OpenAM server directory, default `node["tomcat-openam"]["dns"]`
* `node["tomcat-openam"]["cfg"]["directory"]["port"]` - OpenAM directory port, default `50389`
* `node["tomcat-openam"]["cfg"]["directory"]["admin-port"]` - OpenAM directory admin-port, default `5444`
* `node["tomcat-openam"]["cfg"]["directory"]["jmx-port"]` - OpenAM directory jmx-port, default `5689`
* `node["tomcat-openam"]["cfg"]["root-suffix"]` OpenAM root-suffix, default `o=openam`
* `node["tomcat-openam"]["cfg"]["ds-dirmgrdn"]` - Directory Manager name, default `cn=Directory Manager`
* `node["tomcat-openam"]["cfg"]["ds-dirmgrpasswd"]` - Directory manager password, default `emdstor3me`
* `node["tomcat-openam"]["cfg"]["user-store"]["type"]` - User store type, default `LDAPv3ForOpenDS`
* `node["tomcat-openam"]["cfg"]["user-store"]["ssl"]` - User store SSL, default `SIMPLE`
* `node["tomcat-openam"]["cfg"]["user-store"]["host"]` - User store host (OpenDJ URL) `Only if` you need seperate OpenDJ (for the recipe tomcat-openam::configure)
* `node["tomcat-openam"]["cfg"]["user-store"]["port"]` - User store port, default `1389`
* `node["tomcat-openam"]["cfg"]["user-store"]["suffix"]` - User store suffix, default `dc=example,dc=com`
* `node["tomcat-openam"]["cfg"]["user-store"]["mgrdn"]` - User store manager domain name, default `cn=Directory Manager`
* `node["tomcat-openam"]["cfg"]["user-store"]["passwd"]` - User store password, default `secret12`

Usage
=====

Simply include the recipe where you want Tomcat on apache server with OpenAM installed. This cookbook contains three different recipes, namely

	* full_stack
	* vanilla
	* configure

This cookbook creates a new tomcat server by using our packaged tar ball stored in `S3`. This tar ball has changes to run tomcat behind apache2.

The following changes are needed in the apache2 cookbook. The tomcat tar bundle `server.xml` is equipped to handle the change.

1. By default apache2 cookbook's mod_proxy_http.rb recipe contain the following code

	apache_module "proxy_http"

You just need to change it as follows

	apache_module "proxy_http" do
	  conf true
	end

2. Create a file in /templates/default/mods/`proxy_http.conf.erb` with the following content

	<IfModule mod_proxy_http.c>
	  ProxyRequests Off
	ProxyPreserveHost On
	 
	<Proxy *>
	    Order deny,allow
	    Allow from all
	</Proxy>
	 
	ProxyPass /openam http://localhost:8081/openam
	ProxyPassReverse /openam http://localhost:8081/openam
	<Location /openam>
	    Order allow,deny
	    Allow from all
	</Location>
	</IfModule>

full_stack
----------
This recipe installs apache2, java, tomcat7, OpenAM and OpenDJ in a single instance and the configurations of all has been done in the same instance.

An example for a tomcat-openam full_stack role:

	`name "openam_fullstack"`
	`run_list "recipe[apt]", "recipe[apache2]", "recipe[tomcat-openam::full_stack]"`

vanilla
-------
This recipe installs apache2, java, tomcat7 and OpenAM in an instance. apache2 and tomcat7 has been configured in the same instance. But configuration of OpenAM needs OpenDJ.

An example for a tomcat-openam vanilla role:

	`name "openam_vanilla"`
	`run_list "recipe[apt]", "recipe[apache2]", "recipe[tomcat-openam::vanilla]"`

configure
---------
This recipe is just to configure OpenAM. For this you need OpenDJ. There is a cookbook for OpenDJ in cookbook website as opendj. Run opendj in another instance using the `opendj` cookbook. Give the url id of opendj to `node["tomcat-openam"]["cfg"]["user-store"]["host"]`

An example for a opendj role:

	`name "opendj"`
	`run_list "recipe[apt]", "recipe[opendj::single_instance]"`

An example for a tomcat-openam configure role:

	`name "openam_configure"`
	`run_list "recipe[tomcat-openam::configure]"`


License and Author
==================

Author:: Kishore Kumar (<nkishore@megam.co.in>)
Author:: Thomas Alrin (<alrin@megam.co.in>)


Copyright:: 2013, Megam Systems

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
## v0.1.0:

* Initial version



recipes/vanilla.rb
#
# Cookbook Name:: tomcat-openam
# Recipe:: not_configure_openam
#
# Copyright 2012, Megam Systems
#
# All rights reserved - Do Not Redistribute
#

include_recipe "apache2"
include_recipe "apache2::mod_proxy"
include_recipe "apache2::mod_proxy_http"

case node[:platform]
when "debian", "ubuntu"
  
package "openjdk-7-jre" do
        action :install
end

directory node["tomcat-openam"]["dir-path"]["tmp"] do
  owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  mode node["tomcat-openam"]["mode"]
  action :create
end

remote_file node["tomcat-openam"]["remote-location"]["tomcat-tar"] do
  source node["tomcat-openam"]["source"]["tomcat"]
  mode node["tomcat-openam"]["mode"]
   owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  checksum node["tomcat-openam"]["checksum"] 
end

execute "unzip tomcat" do
  cwd node["tomcat-openam"]["home"]  
  user node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  command node["tomcat-openam"]["cmd"] ["tomcat-unzip"]
end

template node["tomcat-openam"]["remote-location"]["tomcat-init"] do
  source node["tomcat-openam"]["template"]["tomcat_init"]
  owner node["tomcat-openam"]["super-user"]
  group node["tomcat-openam"]["super-user"]
  mode node["tomcat-openam"]["mode"]
end

execute "update tomcat defaults" do
  cwd node["tomcat-openam"]["home"]  
  user node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  command node["tomcat-openam"]["cmd"] ["tomcat-update"]
end



execute "Start tomcat" do
  cwd node["tomcat-openam"]["home"]  
  user node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  command node["tomcat-openam"]["cmd"] ["tomcat-start"]
end


remote_file node["tomcat-openam"]["remote-location"]["openam-tar"] do
  source node["tomcat-openam"]["source"]["openam-war"]
  mode node["tomcat-openam"]["mode"]
  owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  checksum node["tomcat-openam"]["checksum"] 
end

end




recipes/full_stack.rb
#
# Cookbook Name:: tomcat-openam
# Recipe:: full_stack
#It can give an instance with opendj+openam
# Copyright 2012, Megam Systems
#
# All rights reserved - Do Not Redistribute
#
=begin
	This recipe will run an instance with tomcat on apache server. It contains openam with opendj
DEFINITION OF THE BELOW CODE : 

	Refer http://wiki.opscode.com/display/chef/Resources for details

 * Include the apache2 recipe for proxy server
 * Install java with the package openjdk-7-jre
 * Create a temporary directory with the name ~/tmp
 * Download tomcat.tar from our s3 to this insatance
 * Unzip the tomcat tar and initialize it with our template file
 * Update tomcat defaults and start tomcat
 * Download openam.war from our s3 and place it in tomcat/webapps
 * Create a direcory as tmp/openam
 * Download ssoConfiguratorTools.zip into tha above specified folder tmp/openam and unzip it
 * Download opendj.zip into tmp directory and unzip it
 * Move the template file full_stack_cli_config.properties.erb into tmp/openam_cli_config.properties
 * Execute the command to configure opendj
 * Create a directory as home/ubuntu/.openssocfg and home/ubuntu/openam
 * Execute the command to configure openam
 * And finally restart the tomcat
=end

include_recipe "apache2"
include_recipe "apache2::mod_proxy"
include_recipe "apache2::mod_proxy_http"

case node[:platform]
when "debian", "ubuntu"
  
package "openjdk-7-jre" do
        action :install
end

directory node["tomcat-openam"]["dir-path"]["tmp"] do
  owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  mode node["tomcat-openam"]["mode"]
  action :create
end

remote_file node["tomcat-openam"]["remote-location"]["tomcat-tar"] do
  source node["tomcat-openam"]["source"]["tomcat"]
  mode node["tomcat-openam"]["mode"]
   owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  checksum node["tomcat-openam"]["checksum"] 
end

execute "unzip tomcat" do
  cwd node["tomcat-openam"]["home"]  
  user node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  command node["tomcat-openam"]["cmd"] ["tomcat-unzip"]
end

template node["tomcat-openam"]["remote-location"]["tomcat-init"] do
  source node["tomcat-openam"]["template"]["tomcat_init"]
  owner node["tomcat-openam"]["super-user"]
  group node["tomcat-openam"]["super-user"]
  mode node["tomcat-openam"]["mode"]
end

execute "update tomcat defaults" do
  cwd node["tomcat-openam"]["home"]  
  user node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  command node["tomcat-openam"]["cmd"] ["tomcat-update"]
end

execute "Start tomcat" do
  cwd node["tomcat-openam"]["home"]  
  user node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  command node["tomcat-openam"]["cmd"] ["tomcat-start"]
end

remote_file node["tomcat-openam"]["remote-location"]["openam-tar"] do
  source node["tomcat-openam"]["source"]["openam-war"]
  mode node["tomcat-openam"]["mode"]
  owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  checksum node["tomcat-openam"]["checksum"] 
end

directory node["tomcat-openam"]["dir-path"]["tmp-openam"] do
  owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  mode node["tomcat-openam"]["mode"]
  action :create
end

remote_file node["tomcat-openam"]["remote-location"]["ssoConfigurator"] do
  source node["tomcat-openam"]["source"]["ssoconfig-zip"]
  mode node["tomcat-openam"]["mode"]
  owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  checksum node["tomcat-openam"]["checksum"] 
end

package 'zip'
execute "unzip openam ssoconfigurator" do
  command <<CMD
umask 022
unzip -u -o "/home/ubuntu/tmp/openam/ssoConfiguratorTools.zip"
CMD
  cwd node["tomcat-openam"]["dir-path"]["tmp-openam"]
  user node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  action :run
  not_if { ::File.exist?("/home/ubuntu/tmp/openam/configurator.jar") }
end

remote_file node["tomcat-openam"]["remote-location"]["opendj-zip"] do
  source node["tomcat-openam"]["source"]["opendj-zip"]
  mode node["tomcat-openam"]["mode"]
  owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  checksum node["tomcat-openam"]["checksum"] 
end

template node["tomcat-openam"]["remote-location"]["openam-config"] do
  source node["tomcat-openam"]["template"]["openam-full-config"]
  owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  mode node["tomcat-openam"]["mode"]
end

package 'zip'
execute "unzip openam opendj" do
  command <<CMD
umask 022
unzip -u -o "/home/ubuntu/tmp/opendj.zip"

CMD
  cwd "/home/ubuntu"
  user node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  action :run
  not_if { ::File.exist?("/home/ubuntu/opendj/setup") }
end

execute "configurator-opendj" do
  cwd node["tomcat-openam"]["home"]  
  user node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  command node["tomcat-openam"]["cmd"]["config"]["opendj"]
end  

directory node["tomcat-openam"]["dir-path"][".openssocfg"] do
  owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  mode node["tomcat-openam"]["mode"]
  action :create
end

directory node["tomcat-openam"]["dir-path"]["openam"] do
  owner node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  mode node["tomcat-openam"]["mode"]
  action :create
end

execute "configuring-openam" do
  cwd node["tomcat-openam"]["dir-path"][".openssocfg"]
  user node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  command node["tomcat-openam"]["cmd"]["config-sso"]
   action :nothing
end

execute "Restart tomcat" do
  cwd node["tomcat-openam"]["home"]  
  user node["tomcat-openam"]["user"]
  group node["tomcat-openam"]["user"]
  command node["tomcat-openam"]["cmd"] ["tomcat-restart"]
  notifies :run, resources(:execute => "configuring-openam")
end

end




attributes/default.rb

default["tomcat-openam"]["user"] = "ubuntu"
default["tomcat-openam"]["super-user"] = "root"
default["tomcat-openam"]["group"] = "ubuntu"
default["tomcat-openam"]["mode"] = "0755"
default["tomcat-openam"]["checksum"] = "08da002l"

#directory paths
default["tomcat-openam"]["home"] = "/home/ubuntu"
default["tomcat-openam"]["tomcat-home"] = "/home/ubuntu/tomcat"
default["tomcat-openam"]["dir-path"]["webapps-war"] = "/var/lib/tomcat7/webapps/openam.war"
default["tomcat-openam"]["dir-path"]["tmp"] = "/home/ubuntu/tmp"
default["tomcat-openam"]["dir-path"]["openam"] = "/home/ubuntu/openam"
default["tomcat-openam"]["dir-path"]["tmp-openam"] = "/home/ubuntu/tmp/openam"
default["tomcat-openam"]["dir-path"][".openssocfg"] = "/home/ubuntu/.openssocfg"

#Source files
default["tomcat-openam"]["source"]["tomcat"] = "https://s3-ap-southeast-1.amazonaws.com/megam/chef/tomcat/tomcat.tar.gz"
default["tomcat-openam"]["source"]["openam-war"] = "https://s3-ap-southeast-1.amazonaws.com/megam/chef/openam/openam.war"#version 10.0.1
default["tomcat-openam"]["source"]["ssoconfig-zip"] = "https://s3-ap-southeast-1.amazonaws.com/megam/chef/openam/ssoConfiguratorTools.zip"
default["tomcat-openam"]["source"]["opendj-zip"] = "https://s3-ap-southeast-1.amazonaws.com/megam/chef/opendj/opendj.zip"

#Remote files location
default["tomcat-openam"]["remote-location"]["tomcat-tar"] = "/home/ubuntu/tmp/tomcat.tar.gz"
default["tomcat-openam"]["remote-location"]["openam-tar"] = "/home/ubuntu/tomcat/webapps/openam.war"
default["tomcat-openam"]["remote-location"]["ssoConfigurator"] = "/home/ubuntu/tmp/openam/ssoConfiguratorTools.zip"
default["tomcat-openam"]["remote-location"]["opendj-zip"] = "/home/ubuntu/tmp/opendj.zip"
default["tomcat-openam"]["remote-location"]["tomcat-init"] = "/etc/init.d/tomcat"
default["tomcat-openam"]["remote-location"]["openam-config"] = "/home/ubuntu/tmp/openam_cli_config.properties"

#Template files
default["tomcat-openam"]["template"]["tomcat_init"] = 'tomcat_init.sh.erb'
default["tomcat-openam"]["template"]["openam-full-config"] = 'full_stack_cli_config.properties.erb'
default["tomcat-openam"]["template"]["openam-config"] = 'openam_cli_config.properties.erb'

#shell commands
default["tomcat-openam"]["cmd"] ["tomcat-unzip"] = "gunzip -c ~/tmp/tomcat.tar.gz | tar xvf -"
default["tomcat-openam"]["cmd"] ["tomcat-update"] = "sudo update-rc.d tomcat defaults"
default["tomcat-openam"]["cmd"] ["tomcat-start"] = "service tomcat start"
default["tomcat-openam"]["cmd"] ["tomcat-restart"] = "service tomcat restart"

#Arguement's values for the OpenDJ Configuration Command
default["tomcat-openam"]["opendj"]["arg-val"]["baseDN"] = "dc=example,dc=com"
default["tomcat-openam"]["opendj"]["arg-val"]["rootUserDN"] = "cn=Directory Manager"
default["tomcat-openam"]["opendj"]["arg-val"]["rootUserPassword"] = "secret12"
default["tomcat-openam"]["opendj"]["arg-val"]["ldapPort"] = "1389"

#Java Options
default["tomcat-openam"]["java-options"] = "-Xms256m -Xmx1024m"

#Configuration Commands
default["tomcat-openam"]["cmd"]["config"]["opendj"] = "./opendj/setup --cli --baseDN  #{node["tomcat-openam"]["opendj"]["arg-val"]["baseDN"]}  --rootUserDN  '#{node["tomcat-openam"]["opendj"]["arg-val"]["rootUserDN"]}' --rootUserPassword  #{node["tomcat-openam"]["opendj"]["arg-val"]["rootUserPassword"]} -h `hostname` --ldapPort #{node["tomcat-openam"]["opendj"]["arg-val"]["ldapPort"]} --no-prompt"

default["tomcat-openam"]["cmd"]["config-sso"] = "java -jar #{node["tomcat-openam"]["java-options"]} /home/ubuntu/tmp/openam/configurator.jar -f /home/ubuntu/tmp/openam_cli_config.properties > /home/ubuntu/tmp/openam.out"

#AWS public DNS
default["tomcat-openam"]["dns"] = "#{node[:ec2][:public_hostname]}"
default["tomcat-openam"]["instance_id"] = "#{node['ec2']['instance_id']}"

#OpenAM_CLI_Config properties
default["tomcat-openam"]["cfg"]["server-url"] = "http://#{node["tomcat-openam"]["dns"]}:8080"
default["tomcat-openam"]["cfg"]["deployment-uri"] = "/openam"
default["tomcat-openam"]["cfg"]["base-dir"] = "#{node["tomcat-openam"]["home"]}/openam"
default["tomcat-openam"]["cfg"]["locale"] = "en_US"
default["tomcat-openam"]["cfg"]["platform-locale"] = "en_US"
default["tomcat-openam"]["cfg"]["admin-pwd"] = "adminp3me"
default["tomcat-openam"]["cfg"]["amldapuserpasswd"] = "adminl3me"
default["tomcat-openam"]["cfg"]["cookie-domain"] = ".#{node["tomcat-openam"]["dns"]}"
default["tomcat-openam"]["cfg"]["data-store"] = "embedded"
default["tomcat-openam"]["cfg"]["directory"]["ssl"] = "SIMPLE"
default["tomcat-openam"]["cfg"]["directory"]["server"] = "#{node["tomcat-openam"]["dns"]}"
default["tomcat-openam"]["cfg"]["directory"]["port"] = "50389"
default["tomcat-openam"]["cfg"]["directory"]["admin-port"] = "5444"
default["tomcat-openam"]["cfg"]["directory"]["jmx-port"] = "5689"
default["tomcat-openam"]["cfg"]["root-suffix"] = "o=openam"
default["tomcat-openam"]["cfg"]["ds-dirmgrdn"] = "cn=Directory Manager"
default["tomcat-openam"]["cfg"]["ds-dirmgrpasswd"] = "emdstor3me"
default["tomcat-openam"]["cfg"]["user-store"]["type"] = "LDAPv3ForOpenDS"
default["tomcat-openam"]["cfg"]["user-store"]["ssl"] = "SIMPLE"
default["tomcat-openam"]["cfg"]["user-store"]["host"] = "ec2-54-251-5-170.ap-southeast-1.compute.amazonaws.com" #opendj DNS
default["tomcat-openam"]["cfg"]["user-store"]["port"] = "1389"
default["tomcat-openam"]["cfg"]["user-store"]["suffix"] = "dc=example,dc=com"
default["tomcat-openam"]["cfg"]["user-store"]["mgrdn"] = "cn=Directory Manager"
default["tomcat-openam"]["cfg"]["user-store"]["passwd"] = "secret12"

